File: packages/domain/src/__tests__/rbac.test.ts
import { describe, expect, it } from 'vitest';
import { ADMIN_ROLES, PERMISSIONS, ROLE_PERMISSIONS, hasPermission, resolveRules, DEFAULT_BUSINESS_RULES } from '../index';

describe('RBAC', () => {
  it('super admin has every permission', () => {
    for (const p of PERMISSIONS) expect(hasPermission('SUPER_ADMIN', p)).toBe(true);
  });

  it('only super admin can update business rules or manage users', () => {
    for (const role of ADMIN_ROLES.filter((r) => r !== 'SUPER_ADMIN')) {
      expect(hasPermission(role, 'rules.update'), role).toBe(false);
      expect(hasPermission(role, 'users.manage'), role).toBe(false);
    }
  });

  it('front desk can support operations but not touch money rules', () => {
    expect(hasPermission('FRONT_DESK', 'bookings.manage')).toBe(true);
    expect(hasPermission('FRONT_DESK', 'access.simulate')).toBe(true);
    expect(hasPermission('FRONT_DESK', 'members.adjust_credits')).toBe(false);
    expect(hasPermission('FRONT_DESK', 'packages.manage')).toBe(false);
    expect(hasPermission('FRONT_DESK', 'reports.financial')).toBe(false);
  });

  it('finance sees money, not operations mutations', () => {
    expect(hasPermission('FINANCE', 'reports.financial')).toBe(true);
    expect(hasPermission('FINANCE', 'refunds.manage')).toBe(true);
    expect(hasPermission('FINANCE', 'sessions.manage')).toBe(false);
    expect(hasPermission('FINANCE', 'attendance.manage')).toBe(false);
  });

  it('coach is scoped to classes and attendance', () => {
    expect(hasPermission('COACH', 'attendance.manage')).toBe(true);
    expect(hasPermission('COACH', 'ledger.view')).toBe(false);
  });

  it('every role permission is a declared permission', () => {
    for (const role of ADMIN_ROLES) {
      for (const p of ROLE_PERMISSIONS[role]) expect(PERMISSIONS).toContain(p);
    }
  });
});

describe('resolveRules', () => {
  it('branch overrides win over defaults', () => {
    const merged = resolveRules(DEFAULT_BUSINESS_RULES, { qrTtlSeconds: 60 });
    expect(merged.qrTtlSeconds).toBe(60);
    expect(merged.antiPassbackMinutes).toBe(DEFAULT_BUSINESS_RULES.antiPassbackMinutes);
  });
  it('null override returns the defaults', () => {
    expect(resolveRules(DEFAULT_BUSINESS_RULES, null)).toEqual(DEFAULT_BUSINESS_RULES);
  });
});

File: apps/backend/internal/domain/rbac_test.go
package domain

import "testing"

func TestSuperAdminHoldsEveryPermission(t *testing.T) {
	for _, p := range Permissions {
		if !HasPermission(RoleSuperAdmin, p) {
			t.Fatalf("SUPER_ADMIN is missing %s", p)
		}
	}
}

func TestHQAdminIsBlockedFromReshapingTheSystem(t *testing.T) {
	// HQ Admin runs the business but cannot create staff accounts or rewrite
	// the business rules; those stay with Super Admin.
	if HasPermission(RoleHQAdmin, PermUsersManage) {
		t.Fatal("HQ_ADMIN must not manage staff accounts")
	}
	if HasPermission(RoleHQAdmin, PermRulesUpdate) {
		t.Fatal("HQ_ADMIN must not change business rules")
	}
	if !HasPermission(RoleHQAdmin, PermRefundsManage) {
		t.Fatal("HQ_ADMIN should still handle refunds")
	}
	if len(PermissionsFor(RoleHQAdmin)) != len(Permissions)-2 {
		t.Fatalf("HQ_ADMIN holds %d permissions, want %d", len(PermissionsFor(RoleHQAdmin)), len(Permissions)-2)
	}
}

func TestSeparationOfDutiesBetweenFinanceAndOperations(t *testing.T) {
	// Money is Finance's; the schedule is not.
	if !HasPermission(RoleFinance, PermRefundsManage) {
		t.Fatal("FINANCE should manage refunds")
	}
	if HasPermission(RoleFinance, PermSessionsManage) {
		t.Fatal("FINANCE must not schedule classes")
	}

	// The floor is Branch Manager's; the money is not.
	if !HasPermission(RoleBranchManager, PermSessionsManage) {
		t.Fatal("BRANCH_MANAGER should schedule classes")
	}
	if HasPermission(RoleBranchManager, PermRefundsManage) {
		t.Fatal("BRANCH_MANAGER must not issue refunds")
	}
	if HasPermission(RoleBranchManager, PermMembersAdjustCredit) {
		t.Fatal("BRANCH_MANAGER must not hand out credits")
	}
}

func TestFrontDeskAndCoachAreTightlyScoped(t *testing.T) {
	allowed := map[Permission]bool{
		PermDashboardView: true, PermMembersView: true, PermLedgerView: true,
		PermOperationsView: true, PermBookingsManage: true, PermAttendanceManage: true,
		PermAccessView: true, PermAccessSimulate: true, PermCommercialView: true,
		PermPaymentsView: true,
		// The desk runs the till and answers "have you got this in a medium",
		// so it sells, and it can look at stock and at a member's tier.
		PermPOSView: true, PermPOSSell: true,
		PermInventoryView: true, PermCRMView: true,
	}
	for _, p := range Permissions {
		if got := HasPermission(RoleFrontDesk, p); got != allowed[p] {
			t.Fatalf("FRONT_DESK %s = %v, want %v", p, got, allowed[p])
		}
	}

	// The lines that matter on the other side of that: taking money is not the
	// same as unwinding a sale, seeing a shelf is not the same as rewriting
	// what is on it, and nobody at the counter signs for a purchase or hands
	// out loyalty points by hand.
	for _, denied := range []Permission{
		PermPOSVoid, PermInventoryManage, PermInventoryCount,
		PermPurchasingView, PermPurchasingApprove, PermPurchasingReceive,
		PermCRMAdjust, PermCRMManage,
	} {
		if HasPermission(RoleFrontDesk, denied) {
			t.Fatalf("FRONT_DESK must not hold %s", denied)
		}
	}

	// A coach sees their classes and marks attendance, nothing more.
	if !HasPermission(RoleCoach, PermAttendanceManage) {
		t.Fatal("COACH should mark attendance")
	}
	if HasPermission(RoleCoach, PermPaymentsView) || HasPermission(RoleCoach, PermBookingsManage) {
		t.Fatal("COACH must not reach payments or bookings")
	}
	// A coach teaches. The till, the stockroom and the purchase ledger are
	// somebody else's job entirely.
	for _, denied := range []Permission{
		PermPOSView, PermPOSSell, PermInventoryView, PermPurchasingView, PermCRMView,
	} {
		if HasPermission(RoleCoach, denied) {
			t.Fatalf("COACH must not hold %s", denied)
		}
	}
}

func TestPermissionsForIsACopyNotTheLiveMatrix(t *testing.T) {
	perms := PermissionsFor(RoleCoach)
	if len(perms) == 0 {
		t.Fatal("COACH should hold some permissions")
	}
	perms[0] = "tampered"
	if RolePermissions[RoleCoach][0] == "tampered" {
		t.Fatal("PermissionsFor must not expose the underlying matrix")
	}
	if len(PermissionsFor("NOT_A_ROLE")) != 0 {
		t.Fatal("an unknown role must hold nothing")
	}
}

func TestResolveRulesMergesBranchOverrides(t *testing.T) {
	base := DefaultBusinessRules()
	ttl := 90
	policy := PolicyFree
	override := &RulesOverride{QRTTLSeconds: &ttl, NoShowPolicy: &policy}

	merged := ResolveRules(base, override)
	if merged.QRTTLSeconds != 90 {
		t.Fatalf("qr ttl = %d, want the branch override 90", merged.QRTTLSeconds)
	}
	if merged.NoShowPolicy != PolicyFree {
		t.Fatalf("no-show policy = %s, want FREE", merged.NoShowPolicy)
	}
	// Untouched rules keep the organization value.
	if merged.AntiPassbackMinutes != base.AntiPassbackMinutes {
		t.Fatalf("anti-passback drifted to %d", merged.AntiPassbackMinutes)
	}
	if got := ResolveRules(base, nil); got != base {
		t.Fatal("a nil override must leave the defaults untouched")
	}
}
